### ISSUE TITLE
Batch normalization in convolution

### ISSUE THREAD

**Issue by peastman** [deepchem/deepchem#651]
I just spent a few days tracking down a problem that ultimately turned out to be caused by the fact that the Conv2D layer applies batch normalization.  This can cause really subtle, hard to identify incorrect behaviors.  In my case, I had an RL model that contained both a Conv2D and a GRU.  When it was executing a rollout, it would take one step at a time (batch size of 1).  Then it would evaluate the loss function on the whole rollout, which meant the batch normalization would produce different values, and hence the sequence of internal states in the GRU would be different.  So effectively it was selecting steps based on one policy but computing the loss as if it had used a different policy.  The only visible sign of the error was that it didn't do a very good job of learning.

Obviously the Conv2D layer needs an option to select the normalization function.  I'll add that.  I also notice that AtomicConvolution applies batch normalization.  Should it have the same option?

Is batch normalization even a good default behavior?  It can help in some cases, but as seen above, it can also produce really subtle errors.  I'm inclined to say the default normalization function should be None.  That's the default for `tf.contrib.layers.conv2d()`.

**Comment by rbharath** [comment]
I'm inclined to turn off batch norm as default too. As you point out there are many subtle batch norm bugs, and having opt-in seems better.

**Comment by lilleswing** [comment]
I'm inclined to turn off batch norm as default as well.

**Comment by peastman** [comment]
#655 addresses this as far as Conv2D is concerned.  I haven't changed AtomicConvolution because I'm less clear on how that works.  Perhaps someone who knows the atomic convolution code can fix that?  Until it's fixed, any RL model that includes both an AtomicConvolution and a GRU won't work correctly.

**Comment by lilleswing** [comment]
@peastman I can add batch norm as a flag to atomic conv. 

The batch norm helps performance but it isn't integral to the atomic conv architecture.

**Comment by rbharath** [comment]
Going to close this issue since I think it's been resolved. Please re-open if necessary.
